sql-data-pipeline/src/main/java/io/cdap/pipeline/sql/app/core/AbstractSQLExecutor.java

```diff
@@ -0,0 +1,402 @@
+
+    // Load the config
+    config = gson.fromJson(getContext().getSpecification().getProperty(CONFIG_NAME), SQLConfig.class);
+    // Load the plugins
+    for (ETLStage stage: config.getStages()) {
+      String node = stage.getName();
+
+      // Instantiate the plugin
+      SQLTransform plugin;
+      try {
+        plugin = getContext().newPluginInstance(node);
+      } catch (InstantiationException e) {
+        // This probably should have been caught at configure time
+        throw new IllegalStateException("Failed to instantiate plugin for stage " + node);
+      }
+      pluginMap.put(node, plugin);
+    }
+    // Initialize the dag
+    if (!config.getConnections().isEmpty()) {
+      dag = new Dag(config.getConnections());
+    } else {
+      dag = null;
+    }
+    // Load connections maps
+    fromNodeMap = gson.fromJson(properties.get(FROM_CONNECTIONS_NAME),
+                                new TypeToken<Map<String, Set<String>>>() { }.getType());
+    toNodeMap = gson.fromJson(properties.get(TO_CONNECTIONS_NAME),
+                              new TypeToken<Map<String, Set<String>>>() { }.getType());
+    // Load the sources and sinks
+    sourceStages = gson.fromJson(properties.get(SOURCES_NAME), sourceStages.getClass());
+    sinkStages = gson.fromJson(properties.get(SINKS_NAME), sinkStages.getClass());
+  }
+
+  /**
+   * Helper method which loads the Calcite JDBC driver from {@link Driver}.
+   */
+  private void loadCalciteJdbcDriver() {
+    // Load the Calcite JDBC driver
+    try {
+      Class.forName(Driver.class.getName());
+    } catch (ClassNotFoundException e) {
+      // This should not happen assuming this app is properly packaged
+      throw new IllegalStateException("Unable to load the Calcite JDBC driver.");
+    }
+  }
+
+  /**
+   * The actions to perform while iterating through all pipeline nodes in topological order.
+   * @param topologicalOrder The topological ordering of the nodes
+   */
+  private void traverseTopologicalOrder() {
+    // Get topological ordering
+    List<String> topologicalOrder;
+    if (dag != null) {
+      topologicalOrder = dag.getTopologicalOrder();
+    } else {
+      topologicalOrder = new ArrayList<>();
+    }
+
+    // Schema and framework config creation
+    SchemaPlus rootSchema = CalciteSchema.createRootSchema(true).plus();
+    final FrameworkConfig builderConfig = Frameworks.newConfigBuilder().defaultSchema(rootSchema).build();
+    // Identifier for temporary table
+    int temporaryTableCounter = 0;
+    // Traverse the topological ordering
+    for (String node: topologicalOrder) {
+      // Get the plugin
+      SQLTransform plugin = pluginMap.get(node);
+
+      // Create the RelBuilder for generating the query
+      final RelBuilder builder = RelBuilder.create(builderConfig);
+
+      // Generate schema from source nodes
+      // Adding the tables directly to the SchemaPlus seems kind of hacky
+      // TODO: Revisit schema generation
```

Comment:
Can you open JIRA for the TODO, and put the JIRA number here we will not forget later

Reply:
Fixed, opened CDAP-16095